### PR TITLE
Speed optimization training from scratch

### PR DESCRIPTION
**Pull request by tholor**
Optimizing speed for training from scratch using: 
- DistributedDataParallel instead of DataParallel
- AMP
- more workers for StreamingDataSilo

Thanks to @abhinavs95 we got quite some interesting insights on the effect of the above actions.
Let's merge some of the changes in his fork into master. 

**Baseline**
Batch_size = 105
Accumulation steps = 9
Effective batch size = 945
Max sequence length = 128
Iterations = 5k
Time taken = 41 mins
Throughput = 7300 steps / hour => 811 effective batches / hour (batch size 945)
Total batches = 500k
Total estimated training time = 616 hours

**DDP + AMP**
Batch_size = 80 * 4
Accumulation steps = 3
Effective batch size = 960
Max sequence length = 128
Iterations = 4.2k
Time taken = 38 mins
Throughput = 6631 steps / hour => 2210 effective batches / hour (batch size 960)
Total batches = 500k*945/960 = 492187
Total estimated training time = 223 hours

**DDP + AMP + num workers=16**
Batch_size = 80 * 4
Accumulation steps = 3
Effective batch size = 960
Max sequence length = 128
Iterations = 4k
Time taken = 26 mins
Throughput = 9231 steps / hour => 3077 effective batches / hour (batch size 960)
Total batches = 500k*945/960 = 492187
Total estimated training time = 160 hours

**DDP only, num workers=16**
Batch_size = 60 * 4
Accumulation steps = 4
Effective batch size = 960
Max sequence length = 128
Iterations = 5k
Time taken = 45 mins
Throughput = 6666 steps / hour => 1667 effective batches / hour (batch size 960)
Total batches = 500k*945/960 = 492187
Total estimated training time = 295 hours 

(everything measured on a p3.8xlarge with 4x V100)

### CODE REVIEW

**Comment by tholor** [comment]
We finally solved the issue of different ranks running out of sync due. 
So missing steps for merging are:  

- [ ] shuffle data every epoch
- [ ] estimate n_train_steps for original BERT style sequence pairs to steer LR schedule
- [ ] fix saving/loading of checkpoints w/ existing directories
- [ ] fix ZeroDivisionError in eval
- [ ] clean up logging

**Comment by tholor** [comment]
Speed impact of all_reduce seems negligible if we are on a single machine:

With all_reduce:
500 steps => 5:15

Without: 
500 steps => 5:14

(measured on a g3.8xlarge with 2x M60)

If this becomes more significant, we could do this sync only in the last phase (e.g. 10%) of an epoch as ranks won't run out of batches before...

File: Dockerfile-GPU [hunk 1]
```diff
@@ -1,11 +1,33 @@
-FROM nvidia/cuda:10.2-base-ubuntu18.04
+#FROM nvidia/cuda:10.2-base-ubuntu18.04
+FROM pytorch/pytorch:1.5-cuda10.1-cudnn7-devel
+
 RUN apt-get update && apt-get install -y python3-pip && apt-get install -y python3-pip && apt-get install -y git
```
Comment (tanaysoni) [inline Dockerfile-GPU:4]: I think Python install might not be required here anymore as PyTorch image should have Python.

File: farm/modeling/optimization.py [hunk 2]
```diff
@@ -263,8 +265,8 @@ def get_scheduler(optimizer, opts):
 
 def calculate_optimization_steps(n_batches, grad_acc_steps, n_epochs, local_rank):
     optimization_steps = int(n_batches / grad_acc_steps) * n_epochs
-    if local_rank != -1:
-        optimization_steps = optimization_steps // torch.distributed.get_world_size()
+    # if local_rank != -1:
```
Comment (tanaysoni) [inline farm/modeling/optimization.py:268]: Do we need this?

Comment (tholor) [reply farm/modeling/optimization.py:268]: No, we can get rid of it since `n_batches` is already taking into account the number of batches for the current rank.

File: Dockerfile-GPU (hunk 1 above)
Comment (tholor) [reply Dockerfile-GPU]: Yep, good point

File: farm/train.py [hunk 3]
```diff
@@ -324,22 +367,26 @@ def train(self):
...
 
-        if self.log_learning_rate:
-            MlLogger.log_metrics({"learning_rate": self.lr_schedule.get_lr()[0]}, step=self.global_step)
-
         if step % self.grad_acc_steps == 0:
-            # TODO We might wanna add gradient clipping here
+            if self.use_amp:
+                torch.nn.utils.clip_grad_norm_(amp.master_params(self.optimizer), self.max_grad_norm)
+            else:
+                torch.nn.utils.clip_grad_norm_(self.model.parameters(), self.max_grad_norm)
```
[the remark this answers is not in the document]
Comment (tholor) [reply farm/train.py]: Yes, all tasks. Resolving this as with your latest commit we can disable by setting the max_grad_norm parameter to None in the Trainer
